### PR TITLE
Enhancement/lazy loading for prompts on the search page 463

### PR DESCRIPTION
**Pull request by sana7044**
Lazy loading for prompts on the search page

### CODE REVIEW

**Review by marclupanc — CHANGES_REQUESTED** [review]
For some reason this isn't fetching the prompts correctly. Please see the video. 
When navigating to serach page it laods first set of prompts, when you scroll down and reach the observer, you load the second set of prompts and here the issue appears. It loads the prompts that had to be loaded in the first set rather than in the second. 

https://vimeo.com/996647380

**Comment by marclupanc** [comment]
There are still some issues with the way it adds prompts to the search page. The general idea of lazy loading prompts in the search page is as follows:

1. When navigating to serach page we fetch initial set of prompts (sorted by the date) + insert advertisments randomly between those prompts. 
2. When user scrolls down and intersects the observer we load another set of prompts, also sorted by date and which shouldn't be inserted randomly into the search page. 
3. Random insertion should only be for the advertisments 

Let me know if I am mistaken 

https://vimeo.com/998625486?share=copy

**Comment by stravid87** [comment]
Related to #515 & #516.

File: src/pages/SearchPage.vue [hunk 1]
```diff
@@ -78,40 +87,106 @@ const computedAdvertises = computed(() => {
...
+}
+
+onMounted(() => {
+  observer.value.focus()
+  initIntersectionObserver()
+
+  const height = pageRef.value.$el.clientHeight
+  const width = pageRef.value.$el.clientWidth
+  const promptFetchCount = computeFetchPromptCount(height, width)
+
```
Comment (marclupanc) [inline src/pages/SearchPage.vue:186]: If you console.log(promptFetchCount) you will se a number that is computed based on your screen size. But when you will open Pinia store you will see that it actually fetches twice the number specified. This probably might be to the loadMore set to true on initial load. And also when you refresh the page with the log of promptFetchCount you will see that each time the number is increased even though the screen size is still the same

Screenshot 2024-09-09 at 16 50 47
Screenshot 2024-09-09 at 16 51 37

Comment (sana7044) [reply src/pages/SearchPage.vue:186]: I will look into it

**Review by marclupanc — CHANGES_REQUESTED** [review]
We are almost there :D 
Few things I noticed: 
1. There is a fetching when you scroll up for some reason. Not sure if it is necessary?!
2. After fetching next elements when scrolling down, there is a glitch that is bothering our eyes 🤣  

https://vimeo.com/1009179020?share=copy

**Comment by sana7044** [comment]
> We are almost there :D Few things I noticed:
> 
> 1. There is a fetching when you scroll up for some reason. Not sure if it is necessary?!
> 2. After fetching next elements when scrolling down, there is a glitch that is bothering our eyes 🤣
> 
> https://vimeo.com/1009179020?share=copy

1. Yes, it is necessary because the user fetches the latest prompts.
2. please describe it in more detail( what type of glitch in scrolling down)

**Comment by marclupanc** [comment]
> > We are almost there :D Few things I noticed:
> > 
> > 1. There is a fetching when you scroll up for some reason. Not sure if it is necessary?!
> > 2. After fetching next elements when scrolling down, there is a glitch that is bothering our eyes 🤣
> > 
> > https://vimeo.com/1009179020?share=copy
> 
> 1. Yes, it is necessary because the user fetches the latest prompts.
> 2. please describe it in more detail( what type of glitch in scrolling down)

1. It's good that you did it but I think currently it is not needed as we will be implementing caching mechanisms so there will be no sense in updating latest prompts each time, as that will be triggeret at the specific time (or action). Keep this implementation for the future. 
2. Can't you see it on video? at 00:14 and forward. When fetching new prompts there is a significant layout shift

**Comment by sana7044** [comment]
I think this issue is completely fixed now, please check it. @marclupanc

**Review by marclupanc — APPROVED** [review]
Nice job. Maybe in the future we might try and use this library for better performance and optimizations
https://www.npmjs.com/package/vue-virtual-scroller

**Pull request merged**